apps/1d_stencil/stencil_stdexec.cpp

```diff
@@ -0,0 +1,167 @@
+bool header = true;  // print csv heading
+Real_t k = 0.5;      // heat transfer coefficient
+Real_t dt = 1.;      // time step
+Real_t dx = 1.;      // grid spacing
+
+///////////////////////////////////////////////////////////////////////////////
+//[stepper_1
+struct stepper {
+
+    // Our operator
+    Real_t heat(Real_t left, Real_t middle, Real_t right, const Real_t k = ::k, const Real_t dt = ::dt,
+                const Real_t dx = ::dx) {
+        return middle + (k * dt / (dx * dx)) * (left - 2 * middle + right);
+    }
+
+    // do all the work on 'size' data points for 'nt' time steps
+    [[nodiscard]] std::vector<Real_t> do_work(const auto& sch, std::size_t size, std::size_t nt) {
+        std::vector<Real_t> current_vec(size);
+        std::vector<Real_t> next_vec(size);
+
+        auto current_ptr = current_vec.data();
+        auto next_ptr = next_vec.data();
+
+        stdexec::sender auto init = stdexec::transfer_just(sch, current_ptr) |
+                                    stdexec::bulk(size, [&](int i, auto& current_ptr) { current_ptr[i] = (Real_t)i; });
+        stdexec::sync_wait(std::move(init));
+
+        for (std::size_t t = 0; t != nt; ++t) {
+            auto sender =
+                stdexec::transfer_just(sch, current_ptr, next_ptr, k, dt, dx, size) |
+                stdexec::bulk(size, [&](int i, auto& current_ptr, auto& next_ptr, auto k, auto dt, auto dx, auto size) {
+                    std::size_t left = (i == 0) ? size - 1 : i - 1;
+                    std::size_t right = (i == size - 1) ? 0 : i + 1;
+                    next_ptr[i] = heat(current_ptr[left], current_ptr[i], current_ptr[right], k, dt, dx);
+                });
+            stdexec::sync_wait(std::move(sender));
+            std::swap(current_ptr, next_ptr);
+        }
+
+        if (nt % 2 == 0) {
+            return current_vec;
+        }
+        return next_vec;
+    }
+};
+
+///////////////////////////////////////////////////////////////////////////////
+int benchmark(args_params_t const& args) {
+    std::uint64_t size = args.size;  // Number of elements.
+    std::uint64_t nt = args.nt;      // Number of steps.
+    std::string sched = args.sch;
+    heq_sch sch(sched);
+    int nthreads = args.nthreads;
+
+    // Create the stepper object
+    stepper step;
+
+    // Measure execution time.
+    Timer timer;
+
+    // Execute nt time steps on size of elements.
+    // launch with appropriate stdexec scheduler
+    std::vector<Real_t> solution;
+    switch (sch()) {
+        case sch_t::CPU:
+            solution = step.do_work(exec::static_thread_pool(nthreads).get_scheduler(), size, nt);
+            break;
+        case sch_t::GPU:
+            solution = step.do_work(nvexec::stream_context().get_scheduler(), size, nt);
+            break;
+        case sch_t::MULTIGPU:
+            solution = step.do_work(nvexec::multi_gpu_stream_context().get_scheduler(), size, nt);
+            break;
+        case sch_t::INVALID:
```

Comment:
this case can be removed as we need to check for the above three schedulers and default if none of them present. Though imo if we build in the check in the `sch_selector` class itself, the code won't reach here anyway.

Reply:
do you think there is a nice way to put this `switch` block in commons to avoid code replication? i can't think of anything on top of my head.

Reply:
nothing on top of my head, since our algorithms entries (1D, 2D, cholesky) are different (after case statement), and each algorithm takes different input parameters. Not sure what uniform methods can we have.

Reply:
No worries. Thank you!